### DISCUSSION TITLE
Increase the partition count limitation

### DISCUSSION THREAD

**Discussion by tiratano** [milvus-io/milvus#21506] [Q&A and General discussion]
According to a document (https://milvus.io/docs/limitations.md#Number-of-resources-in-a-collection), the maximum number of partitions is 4096 in v2.2.x. However, when I tested the maximum number of partitions in v1.1.1, there was no partition count limitation.

Is there any special reason for this change? I'm planning to assign a partition to some user groups and the group count can be several million numbers (technically could be infinite). The total number of entities might be over billions. The reason for partitioning is that the search speed should be far faster if I specify the partition name. Without partitioning, I should search for a huge amounts of useless entities so the latency could be dramatically slow.

Can I unlock the partition limitations in v2.2.2? Or is there another good solution for this situation?

**Comment by yhmo** [comment]
Partition machinery is not a good design in current version 2.2, we plan to refine partition machinery in v2.3: https://wiki.lfaidata.foundation/display/MIL/Feature+plans+for+Milvus+2.3+and+timeline

Currently, it is not recommended to use too many partitions.

**Reply by Basir-mahmood** [reply]
@yhmo Can you illustrate that how much partitions can we use in milvus-2.4, and is it appropriate to utilize maximum number of partitions? 4096  or more  in milvus-2.4

**Reply by yhmo** [reply]
Compared to v2.2/v2.3, the v2.4 can handle more partitions but still has an upper limit. We strongly recommend to control the total number of partitions to be under 10000(maybe you can create more partitions with some powerful machines). Finally, you will see CPU resources exhausted if you continually add more partitions.

**Comment by tiratano** [comment]
@yhmo Thanks for replying. 
That feature is exactly what I want! According to the link, the release date is 2023.1.10. Will it be released as the planning?

By the way, would the maximum partition count increase? or stay 4096?

**Reply by xiaofan-luan** [reply]
@bigsheeper is currently working on it.
Can you explain your scenario  why you have so many partitions?

**Reply by tiratano** [reply]
@xiaofan-luan I have a lot of users and do not need to search for all the user data. ("search" means similarity search.) As an extreme example, when I have 1 million users and each user has a thousand entities, I just want to search for only 1 user's data. Since 1 user's data is relatively small, the search speed would be really fast. 

Splitting all the data into the 4096 partitions could be one of the solutions, but I just want to save each user's data into each user's partition.

When I tested on v1.1.1, there were no partition limits. Could you make it unlimited again? or increase the limits such as 65,536?

@bigsheeper I'm really looking forward to it. Will v2.3.0 be released as the planning? If not, can I get the estimated date?

**Reply by xiaofan-luan** [reply]
> @bigsheeper is currently working on it. Can you explain your scenario why you have so many partitions?

Even if we increase the partition limit it can not be a huge number, say 1 million users.
The reason is each partition will take some memory resources and maintain such data structure is too expensive.

One thing we could do is hash(userID) and put into a partition, then scalar filtering would work on finding topK result per user.

Another optimization we can do if you can insert same user data in a batch or very close time.
Then all user items will almost in one segment and we can easily skip segments with filters.

**Reply by tiratano** [reply]
@xiaofan-luan 
> Even if we increase the partition limit it can not be a huge number, say 1 million users.
> The reason is each partition will take some memory resources and maintain such data structure is too expensive.

I have several TB memories, so I think I can bear the resource limit. If it's the main reason for the partition limit, could you just put it in the configuration options? I'd like to explicitly define the maximum partition number.

> One thing we could do is hash(userID) and put into a partition, then scalar filtering would work on finding topK result per user.

I've realized that only scalar filtering is not a solution to my problem. (scalar filtering has no speed improvement according to my tests). I'll try this method, however, I still want to increase the partition limit. In some cases, 4096 is not enough size for me.

> Another optimization we can do if you can insert same user data in a batch or very close time.
> Then all user items will almost in one segment and we can easily skip segments with filters.

Sometimes it can be, but I cannot guarantee this. 

In sum, could you make an option for the maximum partition number? For example, 4096 is the default maximum size and if the user can bear the limit, they can set the maximum number. Of course, they should test whether they can use that option or not.

**Reply by xiaofan-luan** [reply]
> ould you make an option for the maximum partition number? For example, 4096 is the default maximum size and if the user can bear the limit, they can set the maximum number. Of course, they

We will definitely refine partition and make it a tunable limit in Milvus 2.3(Target release date 2.28 2023)
But still remember too much partition may introduce small files/indexes, memory consumption and not fully tested, For instance:

By default each segment is 512M, but in your case, if each user use 1 partition, then each partition has a least one segment(might only 10M), and it's not able to flush until time limit reached(By default is 1 one day) If you has 100k partition, then both datanode and querynode need 10M * 100K = 1 T memories.

We are trying to improve the behaviour by force flushing segments but again then there will be many small segments/indexes in the system.

**Reply by tiratano** [reply]
@xiaofan-luan Thanks for your detailed example. That's right. Let's say 1 user has only 10M and I have 100k partitions. Then I need 1 T memories. However, if I have 5 data nodes and 5 query nodes, each node just needs 200 GB of memory, right? This scenario is quite acceptable to me and this is not a problem at all. Plus, I'm also aware that for two in-memory replicas, I need to double the numbers. This is also not a problem, either.

By the way, where is the document for the flush time? I do not know what flush time is. I am planning to delete unused data periodically by myself in order to keep the maximum partition count. If I can delete the unused data on a regular basis, I don't need to flush segments. Am I right?

If I don't miss something, I think all I need is just the large partition count. If all of my assumptions are true, could you release an option for the partition count, first? Since I'm a bit in a hurry, I want to test this design as soon as possible. If this option is released soon in something like v2.2.3, that would be really helpful.

**Reply by xiaofan-luan** [reply]
> I think all I need is just the large partition count. If all of my assumptions are true, could you release an option for the partition count, first? Since I'm a bit in a hurry, I want to test this design as soon as possible. If this option is released soon in something like v2.2.3, that would be really helpful.

2.2.3 can't support dynamic add and drop one partition. I thought if you want use partition per user base then you have to do so.

**Reply by tiratano** [reply]
@xiaofan-luan Regarding 2.2.3, okay. I got it. 

> I thought if you want use partition per user base then you have to do so.

Is "do so" mean "hash(userID) and put into a partition" and all of my assumptions are correct? Okay, I understand "hash(userID)" is the best solution to my case as of now.

> flush until time limit reached(By default is 1 one day)

I am wondering about the meaning of this sentence. Is this related to this link - https://milvus.io/docs/product_faq.md#How-does-Milvus-flush-data? When I insert only 1 record, I can search for the data immediately, however, the data is not on the disk so I can lose the data before explicitly calling flush() (e.g: reboot). Did you warn me about this case?

**Comment by huntrist** [comment]
We are experiencing a similar problem. Just wondering what's the suggested solution here. Is partition hashing the only way to go? Maybe use a combination of large number of collections with many partitions per collection can (partially) solve the problem? Or is creating a large amount of collections bad idea? What's the overhead for collections?

**Reply by yhmo** [reply]
> We are experiencing a similar problem. Just wondering what's the suggested solution here. Is partition hashing the only way to go? Maybe use a combination of large number of collections with many partitions per collection can (partially) solve the problem? Or is creating a large amount of collections bad idea? What's the overhead for collections?

A large number of collections will get similar problems: introduce small files/indexes, increase memory consumption(might cause OOM).
To fix it, we are making an enhancement that could bring large impact, it is still in progress in v2.3.
Currently, I think sharing partition is the only way to reduce partition usage.

**Comment by allendongyx** [comment]
@yhmo  I think shared partition is a very good idea.

**Comment by xiaofan-luan** [comment]
@bigsheeper 
Let's discussion about the the hybrid partition idea
